Add regression tests for sandbox pods when no CNI plugins are initialized

nabeelmohamed opened the pull request:
What is the problem which i am trying to solve?

As mentioned in PR #10744, when no CNI plugins are initialized, the pod network teardown within sandbox_run.go gets deferred and fails, causing NetNS removal to get blocked. This results in a failed pod creation that cannot be cleaned up.

Describing the solution 

To address issue #10768, I've added a regression test to sandbox_run_rollback_test.go to ensure that the sandbox creation fails with the expected error message when no CNI plugins are initialized, as identified in PR #10744. @sameersaeed @mikebrow

Commit:
Add regression tests for sandbox pods when no CNI plugins are initialized

Co-authored-by: Sameer <sameer.saeed@live.ca>

Signed-off-by: nabeelmohamed <mohamednabeel321@gmail.com>

sameersaeed commented on integration/sandbox_run_rollback_test.go:
```diff
@@ -318,6 +318,74 @@ func sbserverSandboxInfo(id string) (*criapiv1.PodSandboxStatus, *types.SandboxI
+			sbConfig := PodSandboxConfig(t.Name(), "nocni", WithPodLabels(labels))
+
+			t.Log("Inject CNI plugin initialization failpoint")
+			injectCNIFailpoint(t, sbConfig, map[string]string{
```
can make the following change to resolve the syntax error:
```			
conf := &failpointConf{
		Add: "1*error(no CNI plugins initialized))",
}
injectCNIFailpoint(t, sbConfig, conf)
```

sameersaeed replied:
could you also please run `gofmt -w -s <containerd-path>/integration/sandbox_run_rollback_test.go` after making the necessary changes to resolve the linter error?

sameersaeed commented:
please also note that PR #10770 has been opened to resolve the issue of pod network teardown failures blocking cleanup

nabeelmohamed commented:
i think that error can be fixed by   
t.Log("Inject CNI plugin initialization failpoint")
conf := &failpointConf{
    Add: "1*error(no CNI plugins initialized)",
}
injectCNIFailpoint(t, sbConfig, conf)    
adding these 'failpointconf' struct to this. since passing a map[string]string to the injectCNIFailpoint function, but it's expecting a pointer to a failpointConf struct

mikebrow commented:
Please rebase now that https://github.com/containerd/containerd/pull/10839 has been merged.

mikebrow commented:
@nabeelmohamed can you rebase your test and make it work correctly with the update?

sameersaeed commented:
Hi @nabeelmohamed, I left some suggestions on your most recent commit - please review and let me know if you have any questions!

sameersaeed commented on integration/sandbox_run_rollback_test.go:
```diff
@@ -319,6 +319,77 @@ func sbserverSandboxInfo(id string) (*criapiv1.PodSandboxStatus, *types.SandboxI
+			t.Log("ListPodSandbox with the specific label")
+			l, err := runtimeService.ListPodSandbox(&criapiv1.PodSandboxFilter{LabelSelector: labels})
+			require.NoError(t, err)
+			require.Len(t, l, 1)
```
We want to make sure that the no pods are up if the 'no CNI plugins initialized' error is thrown, so the number of pods returned by ListPodSandbox should be 0, not 1

nabeelmohamed commented:
Hi @sameersaeed 

I’ve made the changes as per your suggestions. Specifically, I’ve:

Removed the detailed PodSandbox state checks.

Updated the test to assert that zero pods are up when the 'no CNI plugins initialized' error is thrown.

Added cleanup logic to remove the pod used for this check.

Included assertions to ensure no pods are up before and after the containerd restart.

Thanks for your guidance. Please let me know if there are any further adjustments needed.

sameersaeed commented:
Hi @nabeelmohamed, thanks for making the requested changes. There are still some minor nitpicks I have with the current code based on my previous suggestions.

[I've made some additional changes on my own branch](https://github.com/sameersaeed/containerd/blob/d1157c7482cbd80eab100dcc3328ddda3d48262d/integration/sandbox_run_rollback_test.go#L324). Could you please merge these changes onto your branch? 

Please also make sure to rebase and run `gofmt` if needed, so that the tests pass. Thank you!

sameersaeed commented:
Thank you @nabeelmohamed! Just seems like the `gofmt` you ran in commit [0cbcd3a](https://github.com/containerd/containerd/pull/10771/commits/0cbcd3a00cca6af4c629b4d736751ef2a3871db5) modified some unexpected files. We would only need to run it on the file we are changing for this test, `sandbox_run_rollback_test.go`

To fix this, could you please first revert to commit [7526ea1](https://github.com/containerd/containerd/pull/10771/commits/7526ea14450d69d0b566455f12e8bf2bfab20309), (i.e. ```git reset --hard 7526ea1```), and then squash the extra 2 commits that you have on this branch (i.e. ```git reset --soft HEAD~2```)?

After doing the above, you can check if you need to run `gofmt` on `sandbox_run_rollback_test.go` by running `gofmt -l integration` from within your containerd directory. If no output is returned, it should be fine to push without any further changes. Otherwise, if you see the test file being outputted, you can run `gofmt -s -w ./integration/sandbox_run_rollback_test.go` to format the file.

sameersaeed commented:
@nabeelmohamed - It seems like the tests are failing due to the current commit message not including a sign-off

Could you please run `git commit -s --amend` and change the commit message to the following, as mentioned in my previous comment?

```
Add regression tests for sandbox pods when no CNI plugins are initialized
Co-authored-by: Mohamed Nabeel <137497525+nabeelmohamed@users.noreply.github.com>
Signed-off-by: Sameer <sameer.saeed@live.ca>
```

Feel free to reach out to me on the Kubernetes / CNCF Slack if you have any questions on this!

sameersaeed commented:
/retest

sameersaeed commented on integration/sandbox_run_rollback_test.go:
```diff
@@ -319,6 +319,41 @@ func sbserverSandboxInfo(id string) (*criapiv1.PodSandboxStatus, *types.SandboxI
+	t.Log("No pods should be up, as CNI failure is expected to clean up the pod that was attempted to be created")
+	l, err := runtimeService.ListPodSandbox(&criapiv1.PodSandboxFilter{LabelSelector: labels})
+	require.NoError(t, err)
+	require.Len(t, l, 0)
```
From looking over again, it would probably be better to change this `require` on L349 to an `assert`, i.e.: 
```assert.Len(t, l, 0)```

With this change, the test will not immediately exit in a failing case, so we can properly cleanup the pod(s) if needed before finishing the test

Please also rebase if necessary  and make sure to use `git commit -s --amend` for the commit message

mikebrow commented:
/ok-to-test

sameersaeed commented:
Hi @nabeelmohamed, thanks for making the requested changes - seems like all the tests are passing as expected now. 

Could you please rebase once again? Please also make sure to set the commit message using `git -s --amend` once again, as was mentioned in my [previous post](https://github.com/containerd/containerd/pull/10771#issuecomment-2469621520)

nabeelmohamed commented:
hey @sameersaeed ,i dont know why all of a sudden there is 5 failing test which is not in the previos check. all i done is rebase with upstream and force pushed with commit amend message.can you guide me....

sameersaeed commented:
Hi @nabeelmohamed, it seems that there were failures in the latest CI run that are not related to the changes from this PR's test code. You can rerun the tests with the `/retest` command if needed. 

Before you do so, could you please use `git commit -s --amend` to set modify the commit message to the following? 
```
Add regression tests for sandbox pods when no CNI plugins are initialized
Co-authored-by: Sameer <sameer.saeed@live.ca>
Signed-off-by: Mohamed Nabeel <137497525+nabeelmohamed@users.noreply.github.com>
```

sameersaeed commented:
Hi @nabeelmohamed, from checking your most recent commit ([5ac091f](https://github.com/containerd/containerd/commit/5ac091ff3b115e1f71d1f9e484fea5d7afde19fd)), it seems that the commit message is still set as the following:
```
Add regression tests for sandbox pods when no CNI plugins are initialized
Co-authored-by: Mohamed Nabeel <137497525+nabeelmohamed@users.noreply.github.com>

Signed-off-by: nabeelmohamed <mohamednabeel321@gmail.com>
```

As mentioned, could you please change it to the following by using `git commit -s --amend` and editing the commit message?
```
Add regression tests for sandbox pods when no CNI plugins are initialized

Co-authored-by: Sameer <sameer.saeed@live.ca>
Signed-off-by: nabeelmohamed <mohamednabeel321@gmail.com>
```

sameersaeed commented:
Hi @nabeelmohamed - sorry for the delay, I didn't receive a notification on your most recent rebase

Could you please rebase once again and tag me or @mikebrow after doing so? Thank you

nabeelmohamed commented:
> Hi @nabeelmohamed - sorry for the delay, I didn't receive a notification on your most recent rebase
> 
> Could you please rebase once again and tag me or @mikebrow after doing so? Thank you

hey @sameersaeed just rebased and force pushed. also I didn't forget to use git commit --amend command . could you please review this

mikebrow reviewed (approved):
LGTM thanks!
